src/openfermion/utils/_sparse_tools.py

```diff
@@ -250,6 +250,68 @@ def jw_number_restrict_operator(operator, n_electrons, n_qubits=None):
     return operator[numpy.ix_(select_indices, select_indices)]
 
 
+def jw_get_ground_states_by_particle_number(sparse_operator, particle_number):
+    """For a Jordan-Wigner encoded operator, compute the lowest eigenvalue and
+    eigenstates at a particular particle number. The operator must conserve
+    particle number.
+
+    Args:
+        sparse_operator: A Jordan-Wigner encoded sparse operator.
+        particle_number: The particle number at which to compute ground states.
+
+    Returns:
+        ground_energy: The lowest eigenvalue of sparse_operator within the
+            eigenspace of the number operator corresponding to particle_number.
+        ground_states: A list of the corresponding eigenstates.
+
+    Warning: The running time of this method is exponential in the number
+        of qubits.
+    """
+    # Check if operator is Hermitian
+    if not is_hermitian(sparse_operator):
+        raise ValueError('sparse_operator must be Hermitian.')
+
+    n_qubits = int(numpy.log2(sparse_operator.shape[0]))
+
+    # Check if operator conserves particle number
+    sparse_num_op = jordan_wigner_sparse(number_operator(n_qubits))
+    com = commutator(sparse_num_op, sparse_operator)
+    if com.nnz:
+        maxval = max(map(abs, com.data))
+        if maxval > EQ_TOLERANCE:
+            raise ValueError('sparse_operator must conserve particle number.')
+
+    # Get the operator restricted to the sector of the desired particle number,
+    # and compute its eigenvalues and eigenvectors
+    restricted_operator = jw_number_restrict_operator(sparse_operator,
+                                                      particle_number,
+                                                      n_qubits)
+    dense_restricted_operator = restricted_operator.toarray()
+    eigvals, eigvecs = numpy.linalg.eigh(dense_restricted_operator)
+
+    # Get the ground energy and initialize list to store the corresponding
+    # ground states
+    ground_energy = sorted(eigvals)[0]
```

Comment:
Perhaps I am missing something but it seems to me that you are sorting the eigvalues but not the vectors. It seems strange that the eigenvalues returned would be in a potentially different order than the vectors to which they belong. Should you perhaps do numpy.argsort and use that mask to resort both the eigenvectors and values?

Reply:
`sorted` actually returns a new list, so this doesn't sort the eigenvalues. However, I just found out that numpy returns the eigenvalues in sorted order anyway.

Reply:
Oh my bad, I think I missed the zero at the end here and thought you were returning a list of eigenvalues.